packages/jni-swig-stub/realm.i

```diff
@@ -196,7 +196,33 @@ typedef jstring realm_string_t;
 // TODO OPTIMIZATION Optimize...maybe port JStringAccessor from realm-java
 //%typemap(jtype) realm_string_t "String"
 //%typemap(jstype) realm_string_t "String"
-%typemap(in) (realm_string_t) "$1 = rlm_str(jenv->GetStringUTFChars($arg,0));"
+
+// Typemap used for passing realm_string_t into the C-API in situations where the string buffer
+// only have to be live across the C-API call
+%typemap(in) realm_string_t (char* buf) {
+    buf = (char*)jenv->GetStringUTFChars($arg,0);
+    $1.size = strlen(buf);
+    $1.data = buf;
+}
+%typemap(freearg) realm_string_t {
+    if ($1.data) jenv->ReleaseStringUTFChars($arg, (const char *)$1.data);
+}
+
+// Typemap used for passing realm_string_t into the C-API in situations where the string buffer
+// needs to be kept alive after returning from C-API call. This will copy the string buffer to the
+// heap and this have to be explicitly freed at a later point.
+// Currently just matching 'realm_string_t string' arguments to match realm_value_t.string = $input
+%typemap(in) realm_string_t string (char* buf) {
+    buf = (char*)jenv->GetStringUTFChars($arg,0);
+    $1.size = strlen(buf);
+    $1.data = (char const *) (new char[$1.size]);
```

Comment:
Who is freeing this heap-based string? I doubt C-API/Core is doing so, they mostly rely on stack-based allocation and move/copy assignment to deal with passed resources ...

Reply:
The problem is that we create the `realm_value_t` struct and assign the `string` field in different operations in `RealmInterop` which means that the c-stack goes out of scope before calling the actual C-API function with the `realm_value_t` so needs to be _alive_ from assignment and across the actual C-API call ... in which the C-API will correctly copy it.